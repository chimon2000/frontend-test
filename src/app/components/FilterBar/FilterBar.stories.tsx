import * as React from "react";
import { storiesOf } from "@storybook/react";
import { withInfo } from "@storybook/addon-info";
import { action } from "@storybook/addon-actions";
import { boolean, text } from "@storybook/addon-knobs";

import { FilterBar } from "./FilterBar";

const stories = storiesOf("Components", module);

stories.add(
  "Filter Bar",
  withInfo({ inline: true })(() => (
    <FilterBar
      prices={["$", "$$", "$$$"]}
      categories={["French", "Italian", "Japanese"]}
      values={{
        prices: ["$$"],
        categories: ["French"],
        open: false
      }}
      onUpdate={action("Update event")}
    />
  ))
);
