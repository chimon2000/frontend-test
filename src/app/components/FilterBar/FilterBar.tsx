import * as React from "react";
import * as _ from "lodash";

import Dropdown from "../Dropdown";
import Checkbox from "../Checkbox";
import OpenToggleInput from "../OpenToggleInput";

export interface ChangeEventObject {
  name: string;
  value: boolean;
}

export interface FilterValues {
  open: boolean;
  prices: Array<string>;
  categories: Array<string>;
}

export interface FilterBarProps {
  categories: Array<string>;
  prices: Array<string>;
  onUpdate: (name: string, changes: ChangeEventObject) => void;
  values: FilterValues;
}

export const FILTER_DEFAULT: FilterValues = {
  prices: [],
  categories: [],
  open: true
};

export const FilterBar: React.FunctionComponent<FilterBarProps> = ({
  categories,
  prices,
  onUpdate,
  values = FILTER_DEFAULT
}) => {
  return (
    <div className="filter-bar">
      <div className="filter-bar__inline-label">Filter By: </div>
      <div className="filter-bar__inline-input">
        <OpenToggleInput
          className="filter-bar__open-toggle-input"
          id="open-toggle"
          name="is-open"
          checked={values.open}
          onChange={value => onUpdate("open", value)}
        />
      </div>
      <div className="filter-bar__inline-input">
        <Dropdown name="Price" className="filter-bar__price-dropdown-input">
          {prices.map((price: string, idx: number) => (
            <Checkbox
              name={price}
              key={`price-${idx}`}
              id={`price-${idx}`}
              label={price}
              onChange={value => onUpdate("price", value)}
              checked={_.indexOf(values.prices, price) > -1}
            />
          ))}
        </Dropdown>
      </div>
      <div className="filter-bar__inline-input">
        <Dropdown
          name="Categories"
          className="filter-bar__category-dropdown-input"
        >
          {categories.map((category: string, idx: number) => (
            <Checkbox
              name={category}
              key={`category-${idx}`}
              id={`category-${idx}`}
              label={category}
              onChange={value => onUpdate("categories", value)}
              checked={_.indexOf(values.categories, category) > -1}
            />
          ))}
        </Dropdown>
      </div>
    </div>
  );
};
