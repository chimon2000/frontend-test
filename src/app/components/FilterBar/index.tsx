export {
  FilterBar as default,
  ChangeEventObject,
  FilterValues,
  FILTER_DEFAULT
} from "./FilterBar";
