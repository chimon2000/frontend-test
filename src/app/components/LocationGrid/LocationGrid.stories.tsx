// import * as React from "react";
// import { storiesOf } from "@storybook/react";
// import { withInfo } from "@storybook/addon-info";
// import { action } from "@storybook/addon-actions";
// import { withKnobs, boolean, select } from "@storybook/addon-knobs";

// import mockData from "../../__mocks__/search";
// import { LocationGrid, LocationData } from "./LocationGrid";

// const stories = storiesOf("Components", module);

// stories.add("Location Grid", () => <LocationGrid data={mockData.search} />);
