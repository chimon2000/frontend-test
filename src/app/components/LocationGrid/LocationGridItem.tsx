import * as React from "react";
import * as _ from "lodash";

import Button from "../Button";
import StarRating from "../StarRating";
import OpenStatus from "../OpenStatus";

export interface CategoryProps {
  title: string;
}
export interface LocationItemProps {
  id: string;
  name: string;
  rating: number;
  is_closed: boolean;
  price: string;
  categories: Array<CategoryProps>;
  photos: Array<string>;
}

export interface LocationGridItemProps {
  className?: string;
  data: LocationItemProps;
}

export const LocationGridItem: React.FunctionComponent<
  LocationGridItemProps
> = ({
  className,
  data: { id, photos, rating, name, is_closed, price, categories }
}) => {
  const firstCategory: { title: string } = _.first(categories);
  const firstPhoto: string = _.first(photos);

  return (
    <div className="location-grid-item">
      <div className={`location-grid-item__image ${className || ""}`}>
        <img src={firstPhoto} alt={name} />
      </div>
      <div className="location-grid-item__title">{name}</div>
      <div className="location-grid-item__rating">
        <StarRating rating={rating} size="small" />
      </div>
      <div className="location-grid-item__metadata grid">
        <div className="col-6">
          <span className="location-grid-item__category">
            {firstCategory.title}
          </span>
          <span>{" â€¢ "}</span>
          <span className="location-grid-item__price">{price}</span>
        </div>
        <div className="location-grid-item__openstatus col-6">
          <OpenStatus size="small" open={!is_closed} />
        </div>
      </div>
      <Button className="location-grid-item__action" block href={`#/${id}`}>
        Learn More
      </Button>
    </div>
  );
};
