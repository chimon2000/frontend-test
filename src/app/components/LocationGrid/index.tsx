export { LocationGrid as default } from "./LocationGrid";
export { LocationGridItem } from "./LocationGridItem";
