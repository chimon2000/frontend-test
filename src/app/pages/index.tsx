import * as React from "react";
import {
  RouteComponentProps,
  withRouter,
  Route,
  Switch
} from "react-router-dom";

import ListView from "./ListView";
import DetailView from "./DetailView";

export type TParams = { id: string };

class Pages extends React.Component<RouteComponentProps<{}>> {
  constructor(props: RouteComponentProps) {
    super(props);
  }
  render() {
    return (
      <div className="app--container">
        <Route
          path="/:id"
          default={true}
          exact={true}
          render={({ match }: RouteComponentProps<TParams>) => {
            return <DetailView id={match.params.id} />;
          }}
        />
        <Route
          path="/"
          {...this.props}
          exact={true}
          render={() => {
            return <ListView />;
          }}
        />
      </div>
    );
  }
}

export default withRouter(Pages);
