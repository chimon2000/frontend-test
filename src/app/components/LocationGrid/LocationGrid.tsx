import * as React from "react";
import * as _ from "lodash";

import { LocationGridItem, LocationItemProps } from "./LocationGridItem";
import { FilterValues } from "../FilterBar/index";

export interface LocationGridProps {
  className?: string;
  list: Array<LocationItemProps>;
  filters: FilterValues;
}

export const filterList = (
  list: Array<LocationItemProps>,
  filters: FilterValues
) => {
  return _.filter(list, (item: LocationItemProps) => {
    const categoryTitles = _.map(item.categories, "title");
    const isPrice = _.indexOf(filters.prices, item.price) > -1;
    const openState = filters.open === !item.is_closed;

    // because we only display the first category in the list,
    // we should only match on the same - because otherwise it looks like
    // searching is broken.
    const isCategory = _.indexOf(filters.categories, categoryTitles[0]) > -1;

    return isCategory && isPrice && openState;
  });
};

export const LocationGrid: React.FunctionComponent<LocationGridProps> = ({
  children,
  className,
  list,
  filters
}) => {
  const filteredList = filterList(list, filters);

  if (!filteredList.length) {
    return (
      <div className="location-grid location-grid--empty">
        <div className="col-12">
          <h3>0 of {list.length} Results Match Your Filter Criteria.</h3>
          <p>Try something else!</p>
        </div>
      </div>
    );
  }
  return (
    <div className="location-grid">
      {filteredList.map((entry: LocationItemProps, idx: number) => (
        <LocationGridItem key={idx} data={entry} />
      ))}
    </div>
  );
};
