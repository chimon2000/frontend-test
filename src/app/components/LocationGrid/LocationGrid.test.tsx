import * as React from "react";
import { shallow } from "enzyme";

import { LocationGridItem } from "./LocationGridItem";
import { FILTER_DEFAULT } from "../FilterBar";
import { LocationGrid, filterList } from "./LocationGrid";
import { deriveCategories, deriveOptions } from "../../pages/ListView";
import mockData from "../../__mocks__/search";

describe("<LocationGridItem />", () => {
  it("doesn't blow up", () => {
    const data = mockData.search.business;
    const wrapper = shallow(<LocationGridItem data={data[0]} />);

    expect(wrapper).toMatchSnapshot();
  });
});

describe("<LocationGrid />", () => {
  it("doesn't blow up", () => {
    const data = mockData.search.business;
    const wrapper = shallow(
      <LocationGrid list={data} filters={FILTER_DEFAULT} />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it("filterList: filters entries based on criteria", () => {
    const data = mockData.search.business;
    const categories = deriveCategories(data);
    const prices = deriveOptions("price", data);
    const filters = {
      open: true,
      categories,
      prices
    };

    const wrapper = shallow(<LocationGrid list={data} filters={filters} />);

    const children = wrapper.find(".location-grid").children();
    expect(children.length).toEqual(data.length);
  });

  it("filterList: should show nothing with empty filters", () => {
    const data = mockData.search.business;

    const wrapper = shallow(
      <LocationGrid list={data} filters={FILTER_DEFAULT} />
    );

    const children = wrapper.find(".location-grid").children();
    expect(children.length).toEqual(1);
    expect(wrapper.find(".location-grid--empty")).toBeDefined();
  });
});
