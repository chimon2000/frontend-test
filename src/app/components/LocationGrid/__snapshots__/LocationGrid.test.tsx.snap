// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LocationGrid /> doesn't blow up 1`] = `
<div
  className="location-grid location-grid--empty"
>
  <div
    className="col-12"
  >
    <h3>
      0 of 
      20
       Results Match Your Filter Criteria.
    </h3>
    <p>
      Try something else!
    </p>
  </div>
</div>
`;

exports[`<LocationGridItem /> doesn't blow up 1`] = `
<div
  className="location-grid-item"
>
  <div
    className="location-grid-item__image "
  >
    <img
      alt="Mon Ami Gabi"
      src="https://s3-media3.fl.yelpcdn.com/bphoto/ja1-oE6RiIWorNzty-A4Eg/o.jpg"
    />
  </div>
  <div
    className="location-grid-item__title"
  >
    Mon Ami Gabi
  </div>
  <div
    className="location-grid-item__rating"
  >
    <Component
      rating={4}
      size="small"
    />
  </div>
  <div
    className="location-grid-item__metadata grid"
  >
    <div
      className="col-6"
    >
      <span
        className="location-grid-item__category"
      >
        French
      </span>
      <span>
         â€¢ 
      </span>
      <span
        className="location-grid-item__price"
      >
        $$
      </span>
    </div>
    <div
      className="location-grid-item__openstatus col-6"
    >
      <Component
        open={true}
        size="small"
      />
    </div>
  </div>
  <Component
    block={true}
    className="location-grid-item__action"
    href="#/4JNXUYY8wbaaDmk3BPzlWw"
  >
    Learn More
  </Component>
</div>
`;
